### PR TITLE
implement user creation by admin

### PR DESCRIPTION
**Pull request by eug3nix**
the new Add User functionality can be accessed using the *Add User* button on admin/users URI
the manually created users are in confirmed state and thus does not need any extra confirmation from the user.

best regards,
Eugene/CommandPrompt Inc.

### CODE REVIEW

**Comment by hennevogel** [comment]
Hey @eug3nix thank you for your contribution! 💐

**Comment by eug3nix** [comment]
Hey, @hennevogel do you mean the ICHAIN auth setting of OSEM? 
Can I use the ENV['OSEM_ICHAIN_ENABLED'] == 'true' to conditionally disable the user management ?
Thanks.

**Comment by hennevogel** [comment]
yes you can :-)

**Comment by eug3nix** [comment]
here you go, @hennevogel I've adjusted the code per your recomendations and added some tests for such cases

File: app/controllers/admin/users_controller.rb [hunk 1]
```diff
@@ -50,8 +66,14 @@ def edit; end
     private
 
     def user_params
-      params.require(:user).permit(:email, :name, :email_public, :biography, :nickname, :affiliation, :is_admin, :username, :login, :is_disabled,
+      # no password attribute in ICHAIN mode
+      if ENV['OSEM_ICHAIN_ENABLED'] == 'true'
+        params.require(:user).permit(:email, :name, :email_public, :biography, :nickname, :affiliation, :is_admin, :username, :login, :is_disabled,
```
Comment (hennevogel) [inline app/controllers/admin/users_controller.rb]: No need to do this all again. You can [merge](http://api.rubyonrails.org/classes/ActionController/Parameters.html#method-i-merge) `:user` into the params when `ENV['OSEM_ICHAIN_ENABLED']` is set.

File: app/controllers/admin/users_controller.rb [hunk 2]
```diff
@@ -50,8 +66,11 @@ def edit; end
     private
 
     def user_params
-      params.require(:user).permit(:email, :name, :email_public, :biography, :nickname, :affiliation, :is_admin, :username, :login, :is_disabled,
-                                   :tshirt, :mobile, :volunteer_experience, :languages, :to_confirm, role_ids: [])
+      permit_base = [:email, :name, :email_public, :biography, :nickname, :affiliation, :is_admin, :username, :login, :is_disabled,
+                     :tshirt, :mobile, :volunteer_experience, :languages, :to_confirm, role_ids: []]
+      # no password for ICHAIN-enabled mode
+      permit_base << :password unless ENV['OSEM_ICHAIN_ENABLED'] == 'true'
```
Comment (ChrisBr) [inline app/controllers/admin/users_controller.rb]: As this is the admin controller I would just add the ``:password`` to the permit without the conditional.

File: app/views/admin/users/index.html.haml [hunk 3]
```diff
@@ -1,10 +1,14 @@
 .row
   .col-md-12
     .page-header
-      %h2
+      %h1
         Users
         - if @users
           = "(#{@users.length})"
+
+        - if can? :create, User and ENV['OSEM_ICHAIN_ENABLED'] != 'true'
```
Comment (ChrisBr) [inline app/views/admin/users/index.html.haml]: ``
if can? :create, User && ENV['OSEM_ICHAIN_ENABLED'] != 'true'
``

File: app/views/admin/users/index.html.haml [hunk 4]
```diff
@@ -1,10 +1,14 @@
...
+          .pull-right
+            =link_to 'Add User', new_admin_user_path, :class => 'button btn btn-default btn-info'
```
Comment (ChrisBr) [inline app/views/admin/users/index.html.haml]:
```
= link_to 'Add User', new_admin_user_path, class: 'button btn btn-default btn-info' 
```

File: spec/controllers/admin/users_controller_spec.rb [hunk 5]
```diff
@@ -50,4 +50,66 @@
...
+        expect(flash[:error]).to match("Creating User` failed: #{user.errors.full_messages.join('. ')}.")
+      end
+
+      it 'does not create new user' do
+        expect do
+          post :create, user: attributes_for(:user)
+        end.not_to change{ Event.count }
+      end
+    end
+    context 'OSEM_ICHAIN_ENABLED is true' do
```
Comment (ChrisBr) [inline spec/controllers/admin/users_controller_spec.rb]: Line break missing

File: spec/controllers/admin/users_controller_spec.rb [hunk 6]
```diff
@@ -50,4 +50,66 @@
...
+      before do
+        stub_const('ENV', ENV.to_hash.merge('OSEM_ICHAIN_ENABLED' => 'true'))
+        # allow_any_instance_of(User).to receive(:save).and_return(false)
```
Comment (ChrisBr) [inline spec/controllers/admin/users_controller_spec.rb]: Please remove the commented code

File: app/views/admin/users/index.html.haml (hunk 4 above)
Comment (eug3nix) [inline app/views/admin/users/index.html.haml]: @ChrisBr we had a discussion on that with @differentreality some time ago, please see this PR https://github.com/openSUSE/osem/pull/1257
long story short - it is nice to have a visually separated primary and secondary actions from the UX perspective, I used the same approach in the https://github.com/openSUSE/osem/pull/1278
If you guys strongly against that, I would not argue and change the btn to the default green, but I think it is a good step toward more user-friendlt UI

**Comment by ChrisBr** [comment]
LGTM! Restarted travis, hopefully it will resolve the timeout failure ...

File: app/views/admin/users/index.html.haml [hunk 7]
```diff
@@ -1,10 +1,14 @@
 .row
   .col-md-12
     .page-header
-      %h2
+      %h1
         Users
         - if @users
           = "(#{@users.length})"
+
+        - if can? :create, User && ENV['OSEM_ICHAIN_ENABLED'] != 'true'
```
Comment (differentreality) [inline app/views/admin/users/index.html.haml]: We should add the rule in abilities

cannot [:new, :create], User if ENV['OSEM_ICHAIN_ENABLED'] == 'true'

**Comment by differentreality** [comment]
@eug3nix can you please rebase?

File: app/views/admin/users/index.html.haml (hunk 7 above)
Comment (eug3nix) [inline app/views/admin/users/index.html.haml]: @differentreality commits squashed, I've also adjusted the code to get rid of the flash vs flash.now problem that we discussed recently (ref #1429)

File: app/views/admin/users/_form.html.haml [hunk 8]
```diff
@@ -1,26 +1,29 @@
...
+        - else
+          = check_box_tag @user.id, @user.id, @user.confirmed?,
+            method: :patch,
+            url: "/admin/users/#{@user.id}/toggle_confirmation?user[to_confirm]=",
+            class: 'switch-checkbox',
+            readonly: true,
+            data: { size: 'small', on_color: 'success', off_color: 'warning', on_text: 'Yes', off_text: 'No' }
     = f.input :is_admin, hint: 'An admin can create a new conference, manage users and make other users admins.'
-    = f.input :name, as: :string
+    = f.input :name, :as => :string
```
Comment (differentreality) [inline app/views/admin/users/_form.html.haml]: We should stick to the new ruby syntax

File: app/controllers/admin/users_controller.rb [hunk 9]
```diff
@@ -6,6 +6,22 @@ def new
...
 
+    def create
+      # there is no usual user creation process when ICHAIN auth is enabled
+      if ENV['OSEM_ICHAIN_ENABLED'] != 'true'
+        @user = User.new(user_params)
+        @user.skip_confirmation!
+        if @user.save
+          redirect_to admin_users_path, notice: 'User successfully created.'
+        else
+          flash.now[:error] = "Creating User` failed: #{@user.errors.full_messages.join('. ')}."
```
Comment (differentreality) [inline app/controllers/admin/users_controller.rb]: There is a ` after User, not sure what it is doing there

**Comment by differentreality** [comment]
Apart from a few instances of old ruby syntax this looks good, @hennevogel @ChrisBr do you also want to take a look?

File: app/controllers/admin/users_controller.rb [hunk 10]
```diff
@@ -6,6 +6,22 @@ def new
...
+        @user = User.new(user_params)
+        @user.skip_confirmation!
+        if @user.save
+          redirect_to admin_users_path, notice: 'User successfully created.'
+        else
+          flash.now[:error] = "Creating User failed: #{@user.errors.full_messages.join('. ')}."
+          render :new
+        end
+      else
+        redirect_to admin_users_path, error: 'Cannot create user with ICHAIN auth enabled.'
```
Comment (differentreality) [inline app/controllers/admin/users_controller.rb]: Cancancan takes care of this with https://github.com/commandprompt/osem-pg/blob/f9a816eed44625683746192c43acd8fc53f92999/app/controllers/admin/users_controller.rb#L3
So this if statement should not be necessary here, as we should only be able to access the action `create`, if the abilities allow that.

If the authorization fails, cancancan raises an error https://github.com/openSUSE/osem/blob/master/app/controllers/application_controller.rb#L46

File: app/views/admin/users/index.html.haml (hunk 7 above)
Comment (differentreality) [inline app/views/admin/users/index.html.haml]: Please make the addition in the models/ability.rb then you don't need `&& ENV['OSEM_ICHAIN_ENABLED'] != 'true'` here or the if statement in the controller admin/users#create.

File: spec/controllers/admin/users_controller_spec.rb [hunk 11]
```diff
@@ -50,4 +50,66 @@
       end
     end
   end
+  describe 'GET #new' do
+    it 'populates an array of users' do
```
Comment (differentreality) [inline spec/controllers/admin/users_controller_spec.rb]: not a very accurate description

**Comment by differentreality** [comment]
@eug3nix please rebase

File: app/models/ability.rb [hunk 12]
```diff
@@ -119,6 +119,9 @@ def signed_in_with_roles(user)
     # for admins
     can :manage, :all if user.is_admin
 
+    # even admin cannot create new users with ICHAIN enabled
+    cannot [:new, :create], user if ENV['OSEM_ICHAIN_ENABLED'] == 'true'
```
Comment (differentreality) [inline app/models/ability.rb]: This needs to be the `User` instead of `user`, as in the User model.

**Comment by differentreality** [comment]
@eug3nix can you please rebase?

**Comment by differentreality** [comment]
Thanks @eug3nix :smile: 
Let's get it in :+1: :tada:

**Pull request merged**